### PR TITLE
Set the datetime value with time 00:00:00 when using the date mode

### PR DESCRIPTION
**Pull request by fansaien**
Set the datetime value with time 00:00:00 when using the date mode in Date picker.

### CODE REVIEW

**Comment by w20k** [comment]
@fansaien could you describe in detail why do you need it? With an example?

**Comment by fansaien** [comment]
Hello @w20k ,
I need to use the datepicker with the date mode, like this config:
```YAML
fields:
    start_date:
        label: Start Date
        type: datepicker
        mode: date
        span: left
        default: now
```

Right now, the "hidden" input value's time is when the user open this control's time, and can not be changed, like this:
```HTML
<!-- Data locker -->
<input type="hidden" name="start_date" id="Form-field-start_date" value="2019-02-04 22:30:50" data-datetime-value="">
```



I think  if the input value's time always can be set 00:00:00 that will be much better , and also it depends on the user's time zone.

For example, my current timezone is UTC - 6, the HTML is this: 
```HTML
<input type="hidden" name="start_date" id="Form-field-start_date" value="2019-02-04 06:00:00" data-datetime-value="">
```
With this value, I can easy save the date in my controller like this:
```PHP
$model->start_at = DateTime::makeCarbon($data['start_date']);
$model->end_at = DateTime::makeCarbon($data['end_date']);
$model->end_at = $model->end_at->addDay(); 
```

**Comment by LukeTowers** [comment]
@w20k note that he means time set to 0 in accordance with the application's timezone (which should always be UTC, but I digress.

**Comment by w20k** [comment]
@fansaien I'm also against changes inside the core. I'd suggest looking at the JS side, where you could define your '0:00:00' time with ease.

Like so, an example:`$("#datepicker_id_or_class").data('datetimepicker').setLocalDate(new Date(year, month, day, 00, 00));`

**Comment by LukeTowers** [comment]
@w20k note that I'm biased in this issue, @fansaien is my coworker and we're working on a project that would utilize this. I would say that the current behavior of sending a time value isn't ideal when the mode is `date` (just like there's an issue with the date being sent when the mode is `time`). My preferred solution to this would probably be to send only the date itself, but that can get tricky to handle in a robust manner considering timezones and dates potentially being stored in a timestamp or datetime schema column.

**Comment by fansaien** [comment]
Hello @w20k  Thank you for your response and suggestion. I really understand it is not easy to modify the core. 

I think that may be a bug, because it stored a wrong time.

For example:
If a user open the date mode datepicker at 2019-02-05 14:00:00 (UTC), and the begin_date will be stored with that value, the end_date will be 2019-02-06 14:00:00. Then others try to find some data in 2019-2-5, the condition will be 2019-02-05 14:00:00 <= x < 2019-02-06 14:00:00. That can not get the right results.

I suggest the mode option looks like this:

- datetime:  Stored the date and time
- date:         Only stored the date, and time is 0
- time:         Only stored the time 

I also found a related issue, that he only needs the time value with datepicker when the mode is time, https://github.com/octobercms/october/issues/3439

**Comment by w20k** [comment]
@LukeTowers, @fansaien:
Here is a small example, why it could cause a lot of issues.

Example:
Timezone is GMT-6 -> 23.01.2019 with time set to 0:00:00
app.timezone is UTC 0 -> 23.01.2019 with time set to 6:00:00

But if your zone is `+`, like mine:
Timezone is GMT+2 -> 23.01.2019 with time set to 0:00:00
app.timezone is UTC 0 -> 22.01.2019 with time set to 22:00:00 (**You will get a different day.**)

I might be wrong If Carbon doesn't convert date when timezone is set.
```
  $backendTimeZone = \Backend\Models\Preference::get('timezone');
  $value->setTimezone($backendTimeZone);
  $value->setTime(0,0,0);
  $value->setTimezone(Config::get('app.timezone'));
```

**Comment by fansaien** [comment]
Hello @w20k ,

UTC 22.01.2019 with time set to 22:00:00 is not the different day. Because we usually compare with the date time not just date.

If I need to get the results from +2 timezone on 2019-01-23 day,
It will search like this: 2019-01-22 22:00:00(UTC) <= x < 2019-01-23 22:00:00(UTC)

2019-01-22 22:00:00(UTC)  = 2019-01-23 00:00:00 (UTC +2)

**Comment by w20k** [comment]
Hello, @fansaien fair enough.

@LukeTowers will re-run tests once more, and PR is good to go. Though I really do sway toward sending only the date. 

Side note: There is only one corner case, can't imagine who would use it, but when your Backen and DB have different time zones.

File: modules/system/assets/ui/js/datepicker.js [hunk 1]
```diff
@@ -316,7 +316,8 @@
         format: null,
         yearRange: 10,
         firstDay: 0,
-        showWeekNumber: false
+        showWeekNumber: false,
+        mode: 'datetime',
```
Comment (w20k) [inline modules/system/assets/ui/js/datepicker.js]: @fansaien could you remove trailing comma.

Comment (fansaien) [reply modules/system/assets/ui/js/datepicker.js]: Sure, I submit that.

**Comment by LukeTowers** [comment]
@w20k we explicitly tell people to not have a different timezone between your DB and app.timezone config, so that shouldn't be a problem.

**Comment by w20k** [comment]
@fansaien Thank a lot for helping ;)
@LukeTowers testing is done. Could be merged.

**Comment by nizzac** [comment]
This pull request is causing an issue for me in the UI where the Backend Preferences timezone is set to (UTC -00:00) Europe/London.

After a discussion with @w20k it seems the issue is related to timezones that have DST. (Daylight Savings Time)

So when I pick a date with the datepicker that is after the date the clocks go forward... the value set on the hidden input is being set as an hour earlier, thus changing the day...

For example, if I choose 1st April 2019.... rather than the value being set to `2019-04-01 00:00:00` it's being set to `2019-03-31 23:00:00`

**Comment by LukeTowers** [comment]
@fansaien can you comment on the above?

**Comment by fansaien** [comment]
Yes, I will do more test and reply to him.

**Comment by fansaien** [comment]
@nizzac  I tested on both my local timezone -6 and your timezone +1 (After March 31)

For my timezone UTC-6, if I choose 2019-04-02, the value is 2019-04-02 06:00:00(UTC). 

```HTML
<input type="hidden" name="start_date" id="Form-field-start_date" value="2019-04-02 06:00:00" data-datetime-value="">
```
For your timezone London, it is UTC + 1.
The HTML is:
```HTML
<input type="hidden" name="start_date" id="Form-field-start_date" value="2019-04-01 23:00:00" data-datetime-value="">
```
I think that's correct, because the value of input element is not your local timezone, it is the UTC timezone value.

You can also try this website : https://www.timeanddate.com/worldclock/converted.html?iso=20190401T000000&p1=136

2019-04-01 00:00:00(London local Timezone)  = 2019-03-31 23:00:00(UTC)

**Comment by nizzac** [comment]
If you save that data to the database and then try reloading the page. The date set on the date picker is now wrong. With Europe/London timezone...

This is for fields that are set to mode: date

Isnt the issue here that you wantes to set times to 00:00:00 for date pickers? There shouldnt be a need to manipulate the time at all? Especially when doing so changes the date picked.

**Comment by fansaien** [comment]
@nizzac 

I tested with Europe/London timezone on mysql date column and timestamp column.

date column: The date picker is wrong, because the time has been lost.
timestamp column: The date picker is still right.

I suggest using timestamp to save the date value. The datepicker is designed to support datetime, date and time. It will post UTC datetime to the server. So, if the database save timestamp if will be fine.

**Comment by nizzac** [comment]
> date column: The date picker is wrong, because the time has been lost.

This is a problem though right? People will use date columns in the database when they want to store only a date.

Also, if you're using a datepicker with `mode: date` you'd expect the date to be set in the database as you've selected it.

I might be misunderstanding how date and time data is handled here but to me the 4th of April 2019 is  4th of April 2019, regardless of what timezone you're in.

**Comment by fansaien** [comment]
@nizzac 

Hello, please think about the start date and end date or events in the calendar, that should be different.

With the date mode, the previous version will save the current time, and this time can not be changed. In the way, the value should be "2019-04-02 08:54:01", "2019-04-02 15:03:23" or whatever else.

I can not just remove the time of the date time value because I do not know others how to use the time value. My solution is just keep the time value correct.

My suggestions are:
1. If you would like the date column, you can deal with the UTC datetime before you save it.
2. Please create a PR with a new mode name, only keep the date in the value.

**Comment by w20k** [comment]
@fansaien might be related to this PR, could you spend few minutes of your time, if it's not - just ignore it. Thanks a lot!
Issue: https://github.com/octobercms/october/issues/4222

**Comment by LukeTowers** [comment]
@fansaien it's not, just a user error, you can ignore that issue.

**Comment by w20k** [comment]
@fansaien, sorry my bad, rushed with assumption ;) not an issue!

**Comment by fansaien** [comment]
@w20k No worries 👍

**Comment by nizzac** [comment]
> Hello, please think about the start date and end date or events in the calendar, that should be different.

If I was saving an event in a calendar that needed start and end times times then I wouldn't use a DATE column type in the database, I would use a data type that handled dates & times such as a timestamp.

> With the date mode, the previous version will save the current time, and this time can not be changed. In the way, the value should be "2019-04-02 08:54:01", "2019-04-02 15:03:23" or whatever else.

If you need to alter the time of an attribute, then you should use mode: datetime picker.

> I can not just remove the time of the date time value because I do not know others how to use the time value. My solution is just keep the time value correct.

Again, you're talking about time, but this issue is relating to the `mode: date` option for the picker. If you wanted to manipulate time, you should use datetime right?

> @fansaien it's not, just a user error, you can ignore that issue.

I don't think this is a user error.
A backend users timezone setting should make a difference to a value if the datepicker mode is set to date only.

**Comment by LukeTowers** [comment]
@fansaien I think we need to find an alternative solution to the original problem. This is causing issues even on the project that it was originally intended for, see the shipments start date field, it changes after saving

**Comment by w20k** [comment]
@nizzac might cause issues, and should probably work 😄
```
// Set both timezones to UTC to disable converting between them
        if (this.ignoreTimezone) {
            this.appTimezone = 'UTC'
            this.timezone = 'UTC'
        }
```

**Comment by nizzac** [comment]
I don't know what you mean

**Comment by fansaien** [comment]
@LukeTowers , the datepicker will post a UTC datetime string to the server, no matter the mode is datetime, date or time. This datetime string will be converted to PHP Carbon/Datetime type.

If just want to post date or time string to the backend, I think should try to add new mode for that.

**Comment by nizzac** [comment]
That doesn't make any sense. Why create a new mode for something that there is already a mode for?
If the mode is just date, or just time - then the timezone shouldn't come into the equation. 

It only needs to be taken into account when the datepicker is in mode "datetime"

I can't think of a situation where you would want to choose just a date with no ability to change the time of the value and want it to be affected by timezones.

**Comment by LukeTowers** [comment]
@fansaien can we go back to the original problem? Why is it necessary to set the time part of the date when in mode `date` to 00:00:00?

**Comment by fansaien** [comment]
@LukeTowers I do not think we can go back. 

The original version will save the time as the user open this datepicker's value. So the time can be any one: 8:34, 13:21...... To set the 0:0:0, we can know the exact date of users's timezone.

**Comment by nizzac** [comment]
Just out of curiosity, is there a reason you cant add some logic in a beforeSave method that sets the time to 00:00:00 for your particular situation?

**Comment by LukeTowers** [comment]
@fansaien that's not what I meant. Why does the time have to be 0:0:0? I'm starting to think that a beforeSave implementation is a better solution for your original issue.

**Comment by fansaien** [comment]
@LukeTowers 0:0:0 is user's timezone value. When user select one date, the js will convert 0:0:0 to UTC and store into the input value, just like  

<input type="hidden" name="start_date" id="Form-field-start_date" value="2019-04-02 06:00">

I think it is the right value. We can handle this UTC value in the server side.

**Comment by Samuell1** [comment]
I think this PR should be revisited i think it feel little off we setting 00:00:00 even for user timezone it should add hours to it then conver to UTC and save 00:00:00 not in reverse order. @LukeTowers

**Comment by LukeTowers** [comment]
@Samuell1 if you could submit a PR with your proposed fix and a real world example of the problem this caused, then that would help out a lot. I hate dealing with timezone issues, it sucks so much.

**Comment by bennothommo** [comment]
@LukeTowers I've just had a look through this thread. I'm with you - I don't believe the time should play any part whatsoever in a `date` mode datepicker widget, and feel it would be better to strip this out completely and assume that the user is doing the sensible thing and using a `DATE` column to store it in the DB.

**Comment by Samuell1** [comment]
@bennothommo Looks like this line is doing funky stuff because it shouldnt be for user 00:00:00 but converted to timezone https://github.com/octobercms/october/pull/4111/files#diff-c204464daa968cdaa706a81562b46c69R144

**Comment by bennothommo** [comment]
Let's continue the conversation in https://github.com/octobercms/october/issues/5072 - @multiwebinc has submitted an issue that is pretty much covering what's being discussed here.

**Comment by multiwebinc** [comment]
@bennothommo I should have done a search first.

**Comment by bennothommo** [comment]
@multiwebinc That's fine, this was buried deep so it might've been difficult to find. Your issue summarises the problem perfectly so we'll use that to track the bug.
